### PR TITLE
utime： fix __32u sec time overflow

### PR DESCRIPTION
**Pull request by kungf**
when i set osd_scrub_min(max)_interval to a very large num, just like 6307200000(20036524*3600), osd still do scrub. finally i find that the utime_t::nsec only __u32, it's max num is 4294967295, and <
6307200000, so scrub sched_time will overflow, and it will be a random time.

related: #20965

Signed-off-by: kungf <yang.wang@easystack.cn>

### CODE REVIEW

**Comment by kungf** [comment]
@liewegas need a review please

File: src/osd/OSD.cc [hunk 1]
```diff
@@ -6911,10 +6911,18 @@ OSDService::ScrubJob::ScrubJob(CephContext* cct,
     double r = rand() / (double)RAND_MAX;
     sched_time +=
       scrub_min_interval * cct->_conf->osd_scrub_interval_randomize_ratio * r;
+    if (sched_time < timestamp) {
```
Comment (tchaikov) [inline src/osd/OSD.cc]: can we update the implementation of `utime_t& operator+=(utime_t& l, double f)` instead? so if the result will overflow, we can just assign `std::limits<uint32_t>::max()` to it?

File: src/include/utime.h [hunk 2]
```diff
@@ -47,6 +47,11 @@ class utime_t {
   bool is_zero() const {
     return (tv.tv_sec == 0) && (tv.tv_nsec == 0);
   }
+
+  void set_max() {
+    tv.tv_sec = UINT32_MAX;
```
Comment (tchaikov) [inline src/include/utime.h]: might want to use `std::limits<decltype(tv.tv_sec)>::max()` instead.

File: src/osd/OSD.cc (hunk 1 above)
Comment (kungf) [reply src/osd/OSD.cc]: right, file changed, help review,please

**Comment by kungf** [comment]
Jenkins retest this please

**Comment by kungf** [comment]
@tchaikov need a review, please

File: src/include/utime.h [hunk 3]
```diff
@@ -433,23 +438,23 @@ class utime_t {
...
 WRITE_CLASS_DENC(utime_t)
 
-
 // arithmetic operators
 inline utime_t operator+(const utime_t& l, const utime_t& r) {
-  return utime_t( l.sec() + r.sec() + (l.nsec()+r.nsec())/1000000000L,
-                  (l.nsec()+r.nsec())%1000000000L );
+  __u64 sec = (__u64)l.sec() + r.sec() + (l.nsec()+r.nsec())/1000000000ul;
+  __u64 nsec = (l.nsec()+r.nsec())%1000000000ul;
+  return utime_t( ut_safe_convert(sec), ut_safe_convert(nsec));
```
Comment (tchaikov) [inline src/include/utime.h]: is the second ut_safe_convert() necessary?

remove the space after `(`.

why you are using different ways to normalize the result in `operator+` and `operator+=`? could you be more consistent ?

File: src/include/utime.h [hunk 4]
```diff
@@ -32,6 +32,10 @@
 // --------
 // utime_t
 
+//from __u64 to __u32
+inline __u32 ut_safe_convert(__u64 t) {
+  return t < std::numeric_limits<uint32_t>::max() ? t : std::numeric_limits<uint32_t>::max();
```
Comment (tchaikov) [inline src/include/utime.h]: i'd suggest rename `ut_safe_convert()` to `cap_to_u32_max()`. also might want to `use std::min()` here.

**Comment by branch-predictor** [comment]
6307200000 is around 200 years - nobody sane is going to set it so high intentionally. While I agree on the fix, I think it's the osd_scrub_min/max_interval that should be capped to reasonable boundaries.

**Comment by kungf** [comment]
@branch-predictor some one who want close scrub may set it, he may not care how may years he set it, but only want set a very large num, then close the scrub through this way.

**Comment by branch-predictor** [comment]
@kungf to disable scrub, you can use noscrub/nodeepscrub cluster flags. 

>> I think it's the osd_scrub_min/max_interval that should be capped to reasonable boundaries.
>
> what do you mean? only cap osd_scrub_min/max_interval in osd.cc?

No, add appriopriate ```.set_min_max``` at https://github.com/ceph/ceph/blob/master/src/common/options.cc#L3013 .

File: src/include/utime.h [hunk 5]
```diff
@@ -32,6 +32,10 @@
 // --------
 // utime_t
 
+//from __u64 to __u32
```
Comment (tchaikov) [inline src/include/utime.h]: @kungf could you drop this comment. it's unnecessary and redundant here.

Comment (kungf) [reply src/include/utime.h]: ok

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2018-04-17_04:17:03-rados-wip-sage3-testing-2018-04-16-2028-distro-basic-smithi/

**Comment by tchaikov** [comment]
@branch-predictor sorry, i missed your comment. yeah, a min-max bound would be better.

**Pull request merged**